File: haskell/anagram/anagram.hs
module Anagram (anagramsFor) where

import Data.Char (toLower)
import Data.List (sort)

anagramsFor :: String -> [String] -> [String]
anagramsFor word =
  filter (\a -> isAnagram word a)

isAnagram :: String -> String -> Bool
isAnagram word1 word2 =
  (\w1 w2 -> w1 /= w2 && sort w1 == sort w2) lword1 lword2
  where
    lowercase = fmap toLower
    lword1 = lowercase word1
    lword2 = lowercase word2

File: clojure/anagram/src/anagram.clj
(ns anagram
  (:require [clojure.string :as string]))

(defn are-anagrams
  [w1 w2] (let [lw1 (string/lower-case w1) lw2 (string/lower-case w2)]
            (and (= (sort lw1) (sort lw2))
                 (not= lw1 lw2))))

(defn anagrams-for
  [word, candidates]
  (filter (partial are-anagrams word) candidates))
